### ISSUE TITLE
Provide an alternative API for props that uses structs instead of named parameters

### ISSUE THREAD

**Issue by Denommus** [reasonml/reason-react#100]
In [react-react](https://github.com/Denommus/react-react), I have an abstraction on creating ReasonReact components. Each component looks like the following:

```reason
module FooComponent = {
    type props = {name: string};
    let vdomS propsS => S.map eq::(fun _ _ => false)
        (fun {name} => <p> (ReasonReact.stringToElement name) </p>) propsS;
    let component = ReasonReact.reducerComponent "FooComponent";
    let make ::name _children => ReasonReact.componentFromSignal component {name} vdomS;
};
```

So, there's always 3 user-provided parameters:

1. a type (which will be used to pass the props to the signal
2. a signal (which will really model how the component will work)
3. a component name (which will be passed to `ReasonReact.reducerComponent`)

If this _could_ be a functor, so I could abstract my API even further to look like this:

```reason
module FooComponent = GenerateComponent {
    let name = "FooComponent";
    type props = {name: string};
    let vdomS propsS => S.map eq::(fun _ _ => false)
        (fun {name} => <p> (ReasonReact.stringToElement name) </p>) propsS;
};
```

And the rest would be "magic".

But `make` can't be generalized like that. Each `make` function that takes different props will have completely different type signatures from each other.

But if JSX could look for a `make_` function that, instead of using named parameters, used structs (optional parameters would be represented by fields with the `option` type, of course), I could implement `GenerateComponent` like the following:

```reason
module type COMPONENT = {
    open ReactFrp.React;
    let name: string;
    type props;
    let vdomS: signal props => signal ReasonReact.reactElement;
};

module GenerateComponent(C: COMPONENT) => {
    let component = ReasonReact.reducerComponent C.name;
    let make_ (props: C.props) _children => ReactReact.componentFromSignal component props C.vdomS;
};
```

Surely this would benefit more libraries than just react-react. Everyone who wants to abstract away details in the creation of a ReasonReact component could benefit from it. So, do you think the proposal makes sense?

**Comment by let-def** [comment]
What about something like that:
```reason
module type COMPONENT = {
    open ReactFrp.React;
    let name: string;
    type props;
    type prop_constructor 'a;
    let lift : (props -> 'k) -> prop_constructor 'k;
    let vdomS: signal props => signal ReasonReact.reactElement;
};

module GenerateComponent(C: COMPONENT) => {
    let component = ReasonReact.reducerComponent C.name;
    let make_ props children = ReactReact.componentFromSignal component props C.vdomS;
    
    let make_ x : C.prop_constructor (children -> result) = C.lift make_ x
};
```

**Comment by let-def** [comment]
Explanation:
- type `props` is the actual state of the component
- type `prop_constructor` prepends the arguments necessary to construct the state
- `lift` is the computation turning a constructor into a state (providing the default values and any computation necessary at state initialization)

**Comment by Denommus** [comment]
@let-def your solution does work, but I think that would effectively make the `C` module passed to `GenerateComponent` more complex than just writing `FooComponent` by hand as in the first snippet.

A possible second approach that I just realized is using lenses instead of structs.

**Comment by let-def** [comment]
```
module FooComponent = GenerateComponent {
    let name = "FooComponent";
    type state = {name: string};
    type props 'k = ?name:string -> 'k;
    let lift k ?(name="John") = k ?name;
    let vdomS propsS => S.map eq::(fun _ _ => false)
        (fun {name} => <p> (ReasonReact.stringToElement name) </p>) propsS;
};
```

**Comment by Denommus** [comment]
@let-def this looks nice for more complex functors, but in this context it just replaces the need of a user-provided `make` by a user-provided `lift`, which is not a lot of gain.

**Comment by jaredly** [comment]
I've found using a render prop fills in where I would want an HOC functor in lots of cases.

example:
```
    <IngredientsFetcher
      api
      render=(fun fetcherState => {
        switch fetcherState {
        | `Initial => showLoading ()
        | `Loaded (recipe, ingredients) => <MainComponent recipe ingredients api id />
        | `Errored err => showError err
        }
      })
    />
```

**Comment by Denommus** [comment]
@jaredly interesting solution! I'm going to investigate it, thanks!

**Comment by phated** [comment]
I believe this is solved by 0.7.0 if you "hand write" components - see https://reasonml.github.io/reason-react/docs/en/components#hand-writing-components

**Comment by rickyvetter** [comment]
Yeah this is definitely. An interesting usecase I didn’t consider with the hooks release, but this should be quite straightforward now.
